modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C

```diff
@@ -61,6 +61,36 @@ PorousFlowAddMaterialJoiner::act()
       {
         const std::string pf_material_type = params.get<std::string>("pf_material_type");
 
+        _block_restricted = params.isParamValid("block");
+        if (_block_restricted &&
+            (pf_material_type == "fluid_properties" || pf_material_type == "relative_permeability"))
+        {
+          // this Material is block-restricted, and we're about to add Joiners for it, so must get
+          // a union of all the blocks for this material type in order to block-restrict the Joiner
+          _blocks.clear();
+          for (auto & mat2 : _problem->getMaterialWarehouse().getObjects())
+          {
+            const InputParameters & params2 = mat2->parameters();
+            if (params2.isParamValid("pf_material_type") &&
+                params2.get<std::string>("pf_material_type") == pf_material_type)
+            {
+              if (!params2.isParamValid("block"))
+              {
+                // a Material of this type is not block-restricted, so the non-block-restricted
+                // Joiner can be added
+                _block_restricted = false;
+                break;
+              }
+              const std::vector<SubdomainName> & bl =
+                  params2.get<std::vector<SubdomainName>>("block");
+              _blocks.insert(_blocks.end(), bl.begin(), bl.end());
+            }
+          }
+          // eliminate any repeated entries in _blocks
+          const std::set<SubdomainName> unique(_blocks.begin(), _blocks.end());
+          _blocks.assign(unique.begin(), unique.end());
+        }
+
```

Comment:
Could `_blocks` be a set - I don't know? Perhaps you can't insert a vector `bl` easily though? I really don't know.

Reply:
done